---
title: "RelSearch version 1.0.0"
subtitle: "User Manual"
author: "Sho Manabe"
date: "`r Sys.Date()`"
output:
  pdf_document:
    latex_engine: xelatex
    number_sections: true
    toc: true
    toc_depth: 3
urlcolor: blue
documentclass: report
fontsize: 12pt
geometry: margin=1in
header-includes:
  - \usepackage{enumitem}
  - \usepackage{fontspec}
  - \setmainfont{Segoe UI}
  - \usepackage{fvextra}
  - \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
  - \parindent=1em
  - \usepackage{amsmath}
  - \renewcommand{\theequation}{\arabic{equation}}
  - \usepackage{unicode-math}
  - \setmathfont{Cambria Math}
mathjax: null
bibliography: RelSearch_references.bib
csl: forensic-science-international-genetics.csl
---

```{r setup, include = FALSE}
#knitr::opts_chunk$set(tidy = TRUE, tidy.opts = list(width.cutoff = 60))

library(knitr)
library(kableExtra)
```

\newpage

# Initial setup

1. Ensure that R (>= 4.4.0) is installed. It is available from the [R Development Core Team website](http://www.R-project.org).

2. Begin an R session.

3. Execute the following command in R to install required packages.

&nbsp;

```{r, eval = FALSE}
install.packages('https://github.com/manabe0322/RelSearch/releases/download/ v1.0.0/RelSearch_1.0.0.zip', repos = NULL, type = 'win.binary')
install.packages(c("shiny", "data.table", "dplyr", "DT", "magrittr", "pedtools", "Rcpp", "ribd", "shinyFeedback", "shinyjs", "shinythemes", "waiter"))
```

\newpage

# Quick guide

1. Execute the following commands in R to launch the software interface.

&nbsp;

```{r eval = FALSE}
library(RelSearch)
RelSearch()
```

&nbsp;

Then, the software interface will open automatically.

\newpage

2. Load files from each **"Browse..."** button.

![](fig_quick_1.jpg)

**Note** 

- RelSearch can analyze data for each marker type separately as well as for all marker types together.

- Victim, reference, and allele frequency databases are required to analyze autosomal STR profiles.

- Victim and reference databases are required to analyze Y-STR profiles.

- Victim and reference databases are required to analyze mtDNA profiles.

- See section [Files](#sec_files) for information on each file.

\newpage

3. Click the **"Analysis"** button.

![](fig_quick_2.jpg)

\newpage

4. After finishing the analysis, the result window will open automatically.

![](fig_quick_3.jpg)

**Note**

- See section [Result](#sec_result) for navigation of the result window.

\newpage

5. Select **Project** $\blacktriangleright$ **Save project**.

![](fig_quick_4.jpg)

\newpage

6. Enter the project name and click the **"Save as"** button.

![](fig_quick_5.jpg)

**Note**

- The saved project can be loaded from **Project** $\blacktriangleright$ **Load project**.

- How to handle projects is described in section [Other functions](#sec_other_func).

\newpage

# Files {#sec_files}

## STR: Victim database

```{r, echo = FALSE}
df_v_auto_ex <- data.frame(SampleName = c("Victim1", "Victim2", "Victim3", "Victim4", "Victim5"), 
                           D3S1358 = c("17", "17", "15", "15", "15"), 
                           D3S1358 = c("17", "", "17", "18", "15"), 
                           vWA = c("17", "18", "17", "15", "18"), 
                           vWA = c("", "18", "18", "18", ""), 
                           D16S539 = c("11", "", "10", "9", "9"), 
                           D16S539 = c("", "", "10", "12", "9"), 
                           CSF1PO = c("12", "10", "10", "11", "10"), 
                           CSF1PO = c("", "11", "12", "12", ""))
names(df_v_auto_ex) <- c("SampleName", "D3S1358", "D3S1358", "vWA", "vWA", "D16S539", "D16S539", "CSF1PO", "CSF1PO")

kable(df_v_auto_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the column "SampleName" and columns for each marker (two columns in each).

- The marker with two empty cells (e.g., D16S539 of the sample "Victim2" in the above table) is ignored when calculating the likelihood ratio.

- The marker with one empty cell (e.g., vWA of the sample "Victim1" in the above table) can be regarded as both homozygote (i.e., no drop-out) and heterozygote with drop-out of one allele when calculating the likelihood ratio.

&nbsp;

## STR: Reference database

```{r, echo = FALSE}
df_r_auto_ex <- data.frame(SampleName = c("Reference1", "Reference2", "Reference3", "Reference3", "Reference4", "Reference5"), 
                           Family = c("Family 1", "Family 2", "Family 3", "Family 3", "Family 4", "Family 4"), 
                           Relationship = c("Father_Son", "Brother_Brother", "Brother_Brother", "Father_Son", "Brother_Brother", "Uncle_Nephew"),
                           D3S1358 = c("17", "16", "16", "16", "15", "15"), 
                           D3S1358 = c("17", "17", "17", "17", "16", "15"), 
                           vWA = c("14", "17", "14", "14", "17", "14"), 
                           vWA = c("17", "17", "18", "18", "", "16"), 
                           D16S539 = c("10", "10", "9", "9", "9", ""), 
                           D16S539 = c("10", "12", "9", "9", "12", ""))
names(df_r_auto_ex) <- c("SampleName", "Family", "Relationship", "D3S1358", "D3S1358", "vWA", "vWA", "D16S539", "D16S539")

kable(df_r_auto_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the columns "SampleName", "Family", "Relationship", and columns for each marker (two columns in each).

- The names of the relationship need to be pre-defined in **Settings** $\blacktriangleright$ **Relationships** (see section [Setting](#sec_setting_rel)).

- If there are multiple missing family members for a reference, add rows for each relationship of the members (e.g., Brother_Brother and Father_Son of the sample "Reference3" in the above table).

- The marker with two empty cells (e.g., D16S539 of the sample "Reference5" in the above table) is ignored when calculating the likelihood ratio.

- The marker with one empty cell (e.g., vWA of the sample "Reference4" in the above table) can be regarded as both homozygote (i.e., no drop-out) and heterozygote with drop-out of one allele when calculating the likelihood ratio.

## STR: Allele frequencies

```{r, echo = FALSE}
df_af_ex <- data.frame(Allele = c("10", "11", "12", "13", "14", "15"), 
                       D3S1358 = c("", "", "6", "3", "79", "1192"), 
                       vWA = c("", "", "", "1", "586", "78"), 
                       D16S539 = c("602", "562", "517", "209", "26", "3"), 
                       CSF1PO = c("648", "620", "1267", "208", "52", "14"), 
                       TPOX = c("100", "1072", "114", "3", "2", ""))

kable(df_af_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the columns "Allele" and columns for each marker (one column in each).

- Allele counts in the population database are entered in the columns for each marker.

&nbsp;

## Y-STR: Victim database

```{r, echo = FALSE}
df_v_y_ex <- data.frame(SampleName = c("Victim1", "Victim2", "Victim3", "Victim4", "Victim5"), 
                        DYS392 = c("11", "13", "11", "11", ""), 
                        DYS518 = c("37", "38", "37", "37", ""), 
                        DYS570 = c("17", "19", "17", "16", "18"), 
                        DYS437 = c("14", "14", "14", "14", ""), 
                        DYS385 = c("13,17", "10,20", "14", "13", "13,17"))

kable(df_v_y_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the column "SampleName" and columns for each marker (one column in each).

- In the marker with more than one allele, each allele must be separated by a comma without any spaces (e.g., DYS385 in the above table).

- The marker with an empty cell (e.g., DYS392 of the sample ‘Victim5’ in the above table) is ignored for the analysis.

&nbsp;

## Y-STR: Reference database

```{r, echo = FALSE}
df_v_y_ex <- data.frame(SampleName = c("Reference1", "Reference2", "Reference3", "Reference3", "Reference4", "Reference5"), 
                        Family = c("Family 1", "Family 2", "Family 3", "Family 3", "Family 4", "Family 4"), 
                        Relationship = c("Father_Son", "Brother_Brother", "Brother_Brother", "Father_Son", "Brother_Brother", "Uncle_Nephew"), 
                        DYS392 = c("13", "11", "11", "11", "11", "14"), 
                        DYS518 = c("38", "37", "38", "38", "38", ""), 
                        DYS570 = c("19", "17", "16", "16", "19", "18"), 
                        DYS437 = c("14", "14", "14", "14", "14", "14"), 
                        DYS385 = c("10,20", "14,17", "11", "11", "13,15", "13,18"))

kable(df_v_y_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the column "SampleName", "Family", "Relationship", and columns for each marker (one column in each).

- The names of the relationship need to be pre-defined in **Settings** $\blacktriangleright$ **Relationships** (see section [Setting](#sec_setting_rel)).

- If there are multiple missing family members for a reference, add rows for each relationship of the members (e.g., Brother_Brother and Father_Son of the sample "Reference3" in the above table).

- In the marker with more than one allele, each allele must be separated by a comma without any spaces (e.g., DYS385 in the above table).

- The marker with an empty cell (e.g., DYS518 of the sample ‘Reference5’ in the above table) is ignored for the analysis.

&nbsp;

## mtDNA: Victim database

```{r, echo = FALSE}
df_v_mt_ex <- data.frame(SampleName = c("Victim1", "Victim2", "Victim3", "Victim4", "Victim5"), 
                         Range = c("73-340 16024-16365", "73-340 16024-16365", "73-167 240-340 16024-16365", "73-265 16024-16284", ""), 
                         Haplotype = c("16183C 16189C 16217C 16311C 73G 263G 309.1C 315.1C", 
                                       "16093C 16114A 16223T 16362C 73G 263G 309.1C 315.1C", 
                                       "16095T 16189C 16223T 16265C 16274A 16362C 73G 263G", 
                                       "16223T 73G 152C 263G", 
                                       ""))

kable(df_v_mt_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the column "SampleName", "Range", and "Haplotype".

- In the 'Range' column, the first and the last positions in each readable sequence must be written with a hyphen between these positions (e.g, 73-340). The readable sequences must be separated by a single blank space.

- In the 'Haplotype' column, the mtDNA types are expressed as the differences to a reference sequence such as the revised Cambridge Reference Sequence [@Andrews1999] (e.g., 73G, 315.1C, and so on). Each mtDNA type must be separated by a single blank space.

- The empty cells in the 'Range' and 'Haplotype' columns (e.g., the sample ‘Victim5’ in the above table) mean that there is no readable sequence.

&nbsp;

## mtDNA: Reference database

```{r, echo = FALSE}
df_r_mt_ex <- data.frame(SampleName = c("Reference1", "Reference2", "Reference3", "Reference3", "Reference4", "Reference5"), 
                         Family = c("Family 1", "Family 2", "Family 3", "Family 3", "Family 4", "Family 4"), 
                         Relationship = c("Mother_Daughter", "Mother_Daughter", "Brother_Brother", "Father_Son", "Sister_Sister", "Aunt_Niece"), 
                         Range = c("73-340 16024-16365", "73-340 16024-16365", "73-340 16024-16365", "73-340 16024-16365", "73-340 16024-16365", ""), 
                         Haplotype = c("16223T 73G 263G 309.1C 315.1C", 
                                       "16172C 16223T 73G 150T 263G", 
                                       "16172C 16189C 16223T 73G 263G", 
                                       "16172C 16189C 16223T 73G 263G", 
                                       "16223T 16362C 73G 263G 315.1C", 
                                       ""))

kable(df_r_mt_ex, format = "latex", booktabs = T, label = NA, linesep = "") %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the column "SampleName", "Family", "Relationship", "Range", and "Haplotype".

- The names of the relationship need to be pre-defined in **Settings** $\blacktriangleright$ **Relationships** (see section [Setting](#sec_setting_rel)).

- If there are multiple missing family members for a reference, add rows for each relationship of the members (e.g., Brother_Brother and Father_Son of the sample "Reference3" in the above table).

- In the 'Range' column, the first and the last positions in each readable sequence must be written with a hyphen between these positions (e.g, 73-340). The readable sequences must be separated by a single blank space.

- In the 'Haplotype' column, the mtDNA types are expressed as the differences to a reference sequence such as the revised Cambridge Reference Sequence [@Andrews1999] (e.g., 73G, 315.1C, and so on). Each mtDNA type must be separated by a single blank space.

- The empty cells in the 'Range' and 'Haplotype' columns (e.g., the sample ‘Reference5’ in the above table) mean that there is no readable sequence.

\newpage

# Setting {#sec_setting}

## Criteria

In RelSearch, there are some criteria to determine whether each victim-reference pair has an assumed relationship or not. For autosomal STR, RelSearch calculates LR values for each victim-reference pair and investigates whether these LR values are greater than an LR threshold or not. For Y-STR, RelSearch analyzes the number of mismatched loci and the total mismatched steps between victim and reference Y-STR profiles to search paternal lineages. For mtDNA, RelSearch investigates the number of mismatched nucleotides between the victim and the reference sequences to search maternal lineages.

To set criteria, select **Setting** $\blacktriangleright$ **Criteria**. Then the criteria setting window will open automatically.

\newpage

The criteria for each marker type is as follows:

**STR**

- Minimum LR: The minimum likelihood ratio for which the victim and reference are estimated to have the assumed relationship.

**Y-STR**

- Maximum number of mismatched loci: The upper limit of the number of mismatched loci between victim and reference profiles for which paternal lineage is not negated.

- Maximum total mutational steps: The upper limit of total mutational steps (i.e., sum of mutational steps of all loci) between victim and reference profiles for which paternal lineage is not negated.

**mtDNA**

- Maximum number of inconsistency: The upper limit of inconcistency between victim and reference profiles for which maternal lineage is not negated.

&nbsp;

After changing the criteria, press the **Save** button to reflect the changes. Press the **Reset** button if you want to return to default values.

\newpage

## Relationship {#sec_setting_rel}

The names of the relationship in each reference profile need to be pre-defined. To define the relationship names, select **Setting** $\blacktriangleright$ **Relationships**. Then the relationship setting window will open automatically. 

\newpage

There are 44 pre-defined victim-reference relationships as the default. The names of these relationships are determined in the order of victim_reference (e.g., Father_Son means that a victim is the father and a reference is the son). You are free to change these names.

Information on the table is as follows.

- **Sex (Victim)**: The biological sex of the victim

- **Sex (Reference)**: The biological sex of the reference

- **Pr(IBD = 2)**: The probability that two alleles are identity by descent (IBD).

- **Pr(IBD = 1)**: The probability that one allele is identity by descent (IBD).

- **Pr(IBD = 0)**: The probability that zero allele is identity by descent (IBD).

- **Paternal lineage**: Information on whether each relationship is paternal lineage or not.

- **Maternal lineage**: Information on whether each relationship is maternal lineage or not.

You can edit, add, and delete the relationships from the left sidebar.

&nbsp;

### Edit

You can edit the relationship names from the **Edit** button.

In the pop-up window, select a relationship from the defined relationships, and enter a new name of the selected relationship. Press the **Save** button to reflect the new name.

\newpage

### Add

You can define a new relationship from the **Add** button. The procedure is as follows.

1. Enter a relationship name.

2. Make a family tree to define the relationship between a victim and a reference. The following is a list of functions to make a family tree.

- **Add a person**: The function to add an unknown person (UK) to the list of persons.

- **Delete a person**: The function to delete an unknown person (UK) from the list of person.

- **Sex**: The function to select a biological sex (male: M or female: F).

- **Father**: The function to select a person who is the father.

- **Mother**: The function to select a person who is the mother.

- **Founder**: The function to select whether the person is a founder of the family tree or not.

&nbsp;

\noindent Here is an example of defining the brother-brother relationship for the victim-reference pair.

\begin{enumerate}[label = 2.\arabic*., left = 1cm]

  \item Press the **Add** button twice to add the untyped father and mother. UK1 and UK2 are assumed to be the father and mother, respectively.

  \item Set the sex of the Victim, Ref, UK1, and UK2 to "M", "M", "M", and "F", respectively.

  \item In the Founder column, select "No" for Victim and Ref, and select "Yes" for UK1 and UK2.

  \item In the Father column, select "UK1" for Victim and Ref.

  \item In the Mother column, select "UK2" for Victim and Ref.

\end{enumerate}

&nbsp;

3. Press the **View family tree** button to check whether the family tree is correctly set up or not.

4. Press the **Save** button.

&nbsp;

\noindent \textbf{Note}

\noindent Information on the biological sex, IBD probabilities, paternal lineage, and maternal lineage in the new relationship is automatically determined based on your-defined family tree.

### Delete

You can delete a defined relationship from the **Delete** button.

In the pop-up window, select a relationship from the defined relationships. Press the **Save** button to delete the selected relationship.

### Reset

If you want to return to default settings, press the **Reset** button. In the pop-up window, press the **Restore default** button.

### Family tree

You can check the family tree of each defined relationship. Select a row from the relationship table and press the **Family tree** button.

\newpage

## Mutation rate {#sec_setting_myu}

Mutation rates of each autosomal STR marker need to be set to calculate the likelihood ratio (LR) for parent-child relationship. The default mutation rates considering the difference between paternal and maternal and mutational steps (-1, +1, -2, and +2) are reported by Morimoto et al [@Morimoto2020].

To set mutation rates, select **Setting** $\blacktriangleright$ **Mutation rates**. Then the setting window for mutation rates will open automatically. 

### Edit

You can edit the mutation rates from the **Edit** button.

In the pop-up window, select a locus and change the mutation rates. Press the **Save** button to reflect the changes.

### Add

You can add a locus from the **Add** button.

In the pop-up window, enter a locus name and enter the mutation rates. Press the **Save** button to add the locus.

### Delete

You can delete a locus from the **Delete** button.

In the pop-up window, select a locus. Press the **Save** button to delete the selected locus.

### Reset

If you want to return to default settings, press the **Reset** button. In the pop-up window, press the **Restore default** button.

## Data storage conditions {#sec_setting_data_storage}

When handling a large number of victim and reference profiles, it may be difficult to store all the results because RelSearch may run out of memory (see ??? for information on computational burden of RelSearch). You can set the data storage conditions. 

To set the conditions, select **Setting** $\blacktriangleright$ **Data storage conditions**. Then the setting window for data storage will open automatically. 

- Minimum LR to be stored: The minimum LR to be stored in the project data. If you want to reduce the file size of the project data, the minimum LR should be increased.

- Maximum number of displayed data: The maximum number of data displayed in the Results tab. If this number is set too high, RelSearch may run out of memory.

Press the **Save** button to reflect the changes. Press the **Reset** button if you want to return to default values.

\newpage

# Result {#sec_result}

This section describes the display of results on the Results tab.

## Summary

When the analysis is complete, summary data is automatically displayed. You can change the data displayed using the buttons in the left sidebar. You can also download the displayed data as a .csv file using the **Downlaod** button.

\noindent \textbf{Default display}

\noindent Default display depends on whether the input data includes autosomal STR data or not.

If the input data includes autosomal STR data, the data that satisfies the criterion of the minimum LR (see section [Setting](#sec_setting)) is displayed.

If the input data is composed of only Y-STR and/or mtDNA data, the data that satisfies the criteria to support paternal/maternal lineage (see section [Setting](#sec_setting)) is displayed.

If the number of data that satisfies the above conditions is more than the upper limit (i.e., the maximum number of displayed data explained in section [Data storage conditions](#sec_setting_data_storage)), the number of data displayed is the upper limit. 

\noindent \textbf{Identified}

\noindent If the data supports that a victim (or a reference) has the assumed relationship with only one reference (or victim), the victim-reference pair is classified as "Identified".

\noindent \textbf{Inconclusive}

\noindent If the data supports that a victim (or a reference) has the assumed relationship with multiple references (or victims), the victim-reference pair is categorized as "Inconclusive".

\noindent \textbf{Excluded by Y-STR or mtDNA}

\noindent The victim-reference pair is excluded if the LR exceeds the minimum LR but the data does not satisfy the criteria to support paternal/maternal lineage when assuming the paternal/maternal lineage relative.

\noindent \textbf{Minimum LR displayed}

\noindent You can change the minimum LR displayed. After setting the value, press the **Apply** button. Please note that the amount of data stored in the project data is limited according to the data storage conditions (see [Data storage conditions](#sec_setting_data_storage)). Even if you set the minimum LR displayed to less than 1, you cannot check the data with LR < 1 as long as the minimum LR to be stored in the project data is set to 1.

## Selected data in detail

The detailed data of a selected victim-reference pair is displayed in **Selected data in detail** tab.

1. Select a row in the summary table.

2. Go to **Selected data in detail** tab.

In **Selected data in detail** tab, you can review and download the analyzed STR, Y-STR, and mtDNA data for the selected victim-reference pair. You can check all members of the family to which the selected reference belongs.

\noindent \textbf{STR tab}

\noindent \textbf{Y-STR tab}

\noindent \textbf{mtDNA tab}

## All candidates

All candidate relatives that relate to the selected victim-reference pair is displayed in **All candidates** tab. This function is useful when the victim-reference pair is classified as "Inconclusive". 

1. Select a row in the summary table.

2. Go to **All candidates** tab.

In **All candidates** tab, you can download the displayed data.

## Analysis conditions

You can review the following analysis conditions in **Analysis conditions** tab.

- **Database**: Input .csv file names of each database are displayed.

- **Allele probability**: Unobserved alleles in the population database are displayed, if any. You can download the allele probabilities that are estimated based on Dirichlet distribution (see section [Computational principle](#sec_principle)).

- **Criteria**: The user-defined criteria to determine whether each victim-reference pair has an assumed relationship or not (see section [Setting](#sec_setting)) is displayed.

- **Assumed relationship**: Information on the user-defined relationships (see section [Setting](#sec_setting_rel)) is displayed.

- **Mutation rate**: The user-defined mutation rates (see section [Setting](#sec_setting_myu)) are displayed.

- **Data storage conditions**: The user-defined conditions for data storage (see section [Setting](#sec_setting_data_storage)) is displayed.

\newpage

# Other functions {#sec_other_func}

## Project

### New project

To start a new project, go to **Project** $\blacktriangleright$ **New project** and press the **New project** button.

**Important**: The current project should be saved before starting a new project because the current project will be deleted.

### Load project

To load a previous project, go to **Project** $\blacktriangleright$ **Load project**. Select a project file from the **Browse...** button and press the **Load project** button.

**Important**: The current project should be saved before loading another project because the current project will be deleted.

### Save project

To save the current project, go to **Project** $\blacktriangleright$ **Save project**. Enter the project name and press the **Save as** button.

## View database

Each loaded database can be reviewed in **Database** tab.

## Example files

Example files for each database can be downloaded from the **Example files** tab.

\newpage

# Computational burden

When handling a large number of victim and reference profiles, the computational load of RelSearch will be high. The following table shows approximate computation time and memory limits as a guide.

```{r, echo = FALSE}
df_comp <- data.frame(n_vic = c("1,000", "2,000", "3,000", "4,000", "5,000"), 
                      n_ref = c("1,000", "2,000", "3,000", "4,000", "5,000"), 
                      n_pair = c("1 million", "4 million", "9 million", "16 million", "25 million"), 
                      marker = c("STR, Y-STR, mtDNA", "", "", "", ""),
                      time = c("3 min", "15 min", "35 min", "", "130 min"))
names(df_comp) <- c("Number of victims", "Number of references", "Number of victim-reference pairs", "Marker type", "Computational time")

kable(df_comp, format = "latex", booktabs = T, label = NA, linesep = "", escape = FALSE) %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

Computer specifications are Windows 10 64 bit, Intel Core i9-9900K CPU, 3.6 GHz, 32 GB RAM. Data storage conditions are as follows:

- Minimum LR to be stored: 1

- Maximum number of displayed data: 10,000

\newpage

# Computational principle {#sec_principle}

## STR

### Likelihood ratio

The likelihood ratio (LR) is calculated for each victim-reference pair by assuming the following two hypotheses:

$H_{1}$: The victim and the reference are a certain relationship,

$H_{2}$: The victim and the reference are unrelated.

The equation of the LR is as follows:

\begin{equation}
LR = \frac{Pr(G_{v}, G_{r}|H_{1})}{Pr(G_{v}, G_{r}|H_{2})}
\end{equation}

\noindent where $G_{v}$ and $G_{r}$ denote the genotypes of the victim and the reference, respectively.

In RelSearch, users can set the minimum threshold of the LR to estimate that the victim and reference have the relationship assumed in $H_{1}$. The default threshold is 100, which is determined based on the verbal scale "Moderate Support" written in the Scientific Working Group on DNA Analysis Methods (SWGDAM) guideline [@SWGDAM2018].

If full profiles are obtained for a victim-reference pair under the assumption of the relationship other than the parent-child (i.e., without considering drop-out and mutations in calculation), the numerator and the denominator of the LR are calculated using the probabilities that two, one and zero alleles are identity by descent (IBD).

\begin{equation}
\begin{aligned}
Pr(G_{v}, G_{r}|H) &= Pr(G_{v}, G_{r}|IBD = 2)Pr(IBD = 2|H)
\\ &+ Pr(G_{v}, G_{r}|IBD = 1)Pr(IBD = 1|H)
\\ &+ Pr(G_{v}, G_{r}|IBD = 0)Pr(IBD = 0|H)
\end{aligned}
\end{equation}

The specific method for these calculations is based on the method of Wenk et al [@Wenk1996].

### Allele probabilities

The allele probabilities are assumed to be based on Dirichlet distributions in each locus [@Morimoto2020]. The probability of allele $x$ ($p_x$) can be estimated using the following formula:

\begin{equation}
p_x = \frac{\alpha_x + 1}{\sum_{x = 1}^X(\alpha_x + 1)}
\end{equation}

where $\alpha_x$ is the number of observations of the $x$th allele ($x = 1, 2, ..., X$) and X is the number of observed allele types.

If some alleles of victims or references are unobserved in the population database, these alleles are added to the observed allele types as zero observation (i.e., $\alpha_x = 0$), and $p_x$ of all alleles are re-estimated [@Morimoto2020].

### Allelic drop-out

When only one allele is observed and another allele may have dropped out in a locus, RelSearch assumes that the genotype is both homozygous (i.e., no drop-out allele) and heterozygous (i.e., one drop-out allele). The probability of observing the profile is the same in all candidate genotypes. The method corresponds to the "Method 1" proposed by Dørum et al [@Dorum2015].

Here, $G_{v}^*$ and $G_{r}^*$ denote the observed profiles of the victim and the reference, respectively. $G_{v_i}$ and $G_{r_j}$ denote the $i$th candidate genotype of the victim and the $j$th candidate genotype of the reference, respectively.

Suppose that $G_{v}^*$ is 15/- (i.e., partial profile) and $G_{r}^*$ is 16/17. In this case, RelSearch considers the following candidate genotypes of the victim:

$G_{v_1}$: 15/15, 

$G_{v_2}$: 15/16, 

$G_{v_3}$: 15/17, and

$G_{v_4}$: 15/Q.

Q denotes any undetected alleles other than 15, 16, and 17.

The candidate genotype of the reference is only 16, 17 because two heterozygote alleles are detected.

In this case, the likelihood function considering allelic drop-out is as follows:

\begin{equation}
\begin{aligned}
&Pr(G_{v}^* : 15/-, G_{r}^* : 16/17|H)
\\ &= Pr(G_{v}^* : 15/-, G_{r}^* : 16/17|G_{v_i} : 15/15, G_{r} : 16/17)Pr(G_{v_i} : 15/15, G_{r} : 16/17|H)
\\ &+ Pr(G_{v}^* : 15/-, G_{r}^* : 16/17|G_{v_i} : 15/16, G_{r} : 16/17)Pr(G_{v_i} : 15/16, G_{r} : 16/17|H)
\\ &+ Pr(G_{v}^* : 15/-, G_{r}^* : 16/17|G_{v_i} : 15/17, G_{r} : 16/17)Pr(G_{v_i} : 15/17, G_{r} : 16/17|H)
\\ &+ Pr(G_{v}^* : 15/-, G_{r}^* : 16/17|G_{v_i} : 15/Q, G_{r} : 16/17)Pr(G_{v_i} : 15/Q, G_{r} : 16/17|H)
\end{aligned}
\end{equation}

This equation is generalized as follows:

\begin{equation}
Pr(G_{v}^*, G_{r}^*|H) = \sum_i \sum_j Pr(G_{v}^*, G_{r}^*|G_{v_i}, G_{r_j})Pr(G_{v_i}, G_{r_j}|H)
\end{equation}

$Pr(G_{v}^*, G_{r}^*|G_{v_i}, G_{r_j})$ is assumed to be the same irrespective of the candidate genotypes. Therefore, the likelihood function can be rewritten as follows:

\begin{equation}
Pr(G_{v}^*, G_{r}^*|H) = \sum_i \sum_j Pr(G_{v_i}, G_{r_j}|H)
\end{equation}

$Pr(G_{v_i}, G_{r_j}|H)$ values are calculated according to equation (1) under the assumption of the relationship other than the parent-child. When assuming the parent-child relationship, mutational events are considered for the calculation as explained in the next session.

### Mutation

When assuming the parent-child relationship, mutational events are considered for the calculation of $Pr(G_{v_i}, G_{r_j}|H)$ in equation (2). Suppose that a victim-reference pair is assumed to be parent-child, and $G_{v_i}$ is a/b and $G_{r_j}$ is c/d. RelSearch assumes all inheritance patterns from the parent to the child including mutational events. The LR is calculated according to [@Ricciardi2014; @Egeland2017].

\begin{equation}
LR = \frac {(\mu_{a \rightarrow c} + \mu_{b \rightarrow c})p_{d} + (\mu_{a \rightarrow d} + \mu_{b \rightarrow d})p_{c}}{4p_{c}p_{d}}
\end{equation}

where $p_a$, $p_b$, $p_c$, and $p_d$ denote the probabilities of the allele $a$, $b$, $c$, and $d$, respectively. $\mu_{a \rightarrow c}$ denotes the probability that the parent allele $a$ is inherited as the allele $c$ to the child. $\mu_{a \rightarrow c}$, $\mu_{b \rightarrow c}$, $\mu_{a \rightarrow d}$, and $\mu_{b \rightarrow d}$ depend on the sex of the parent (i.e., paternal or maternal), and the mutational steps are assumed to be -2, -1, +1, and +2 steps. The mutation rates reported by Morimoto et al [@Morimoto2020] are used as the default values.

\newpage

## Y-STR

RelSearch analyzes the number of mismatched loci and the total mutational steps between victim and reference Y-STR profiles. The following table shows some examples of the analysis results in one locus.

```{r, echo = FALSE}
df_y_interpret_ex <- data.frame(vic_prof = c("15", "15", "15", "-", "15", "15", "15, 16", "15, 16", "15, 17"), 
                                ref_prof = c("15", "16", "17", "15", "15, 17", "16, 17", "15, 17", "15, 18", "16, 18"), 
                                mismatch = c("", "Yes", "Yes", "", "", "Yes", "Yes", "Yes", "Yes"), 
                                mu_step = c("0", "1", "2", "", "", "1", "1", "2", "Unable to calculate"), 
                                ignore_do = c("", "", "", "Yes", "Yes", "", "", "", ""))
names(df_y_interpret_ex) <- c("Victim profile", "Reference profile", "Mismatch", "Mutational steps", "Ignore due to allelic drop-out")

kable(df_y_interpret_ex, format = "latex", booktabs = T, label = NA, linesep = "", escape = FALSE) %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

RelSearch ignores loci where the victim and reference profile match when considering the allelic drop-out. Suppose that the victim profile is 15, and the reference profile is 15, 17 in a duplicated marker (e.g., DYS385). If the allele 17 is assumed to be dropped out in the victim profile, the assumed victim profile is the same as the reference profile.

If multiple mutational events are assumed in a duplicated marker, RelSearch does not calculate the mutational steps for the marker due to the difficulty of the estimation of the actual mutational process (i.e., "Unable to calculate" in the above table). Especially in DYF387S1, three alleles may be observed and the mutational process related to these alleles is too complex.

After analyses of all victim-reference profiles, RelSearch investigates whether each victim-reference profile supports paternal lineage or not based on the following criteria.

- Maximum number of mismatched loci: The upper limit of the number of mismatched loci between victim and reference profiles for which paternal lineage is not negated.

- Maximum total mutational steps: The upper limit of total mutational steps between victim and reference profiles for which paternal lineage is not negated.

The default criteria for the maximum number of mismatched loci is 4 and for the maximum total mutational steps is 5, which is determined based on a previous study [@Liu2016].

\newpage

## mtDNA

RelSearch investigates the number of mismatched nucleotides between the victim and the reference sequences. There is one criterion to support maternal lineage as follow:

- Maximum number of inconsistency: The upper limit of inconcistency between victim and reference profiles for which maternal lineage is not negated.

Information on the ranges of the mtDNA sequences in each profile is needed because RelSearch targets the ranges read in both victim and reference profiles. The mtDNA profiles are expressed as the differences to a reference sequence such as the revised Cambridge Reference Sequence [@Andrews1999] (e.g., 73G, 315.1C, and so on). RelSearch only recongizes differences in strings of each mtDNA profile; therefore, the rule of the notation for mtDNA profiles should be aligned in each profile by software users.

The default maximum number of inconsistency is 1, which is determined according to the SWGDAM guideline for mtDNA analysis interpretation [@SWGDAM2019].

\newpage

# Reference {-}

<div id="refs"></div>
