---
title: "RelSearch version 1.0.0 user manual"
output:
  pdf_document:
    latex_engine: xelatex
    number_sections: true
    toc: true
    toc_depth: 3
urlcolor: blue
documentclass: article
fontsize: 12pt
geometry: margin = 1in
header-includes:
  - \usepackage{indentfirst}
  - \parindent = 1em
bibliography: relsearch_references.bib
csl: forensic-science-international-genetics.csl
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(tidy = TRUE, tidy.opts = list(width.cutoff = 60))
library(knitr)
library(kableExtra)
```

\newpage

# Initial setup

1. Ensure that R (>= 4.4.0) is installed. It is available from the [R Development Core Team website](http://www.R-project.org).

2. Begin an R session.

3. Execute the following command in R to install required packages.

&nbsp;

```{r eval = FALSE}
install.packages(c("shiny", "data.table", "dplyr", "DT", "magrittr", "pandoc", "pedtools", "Rcpp", "ribd", "rmarkdown", "shinyFeedback", "shinyjs", "shinythemes", "waiter"))
```

&nbsp;

4. Download ”relsearch_1.0.0.zip” from the [GitHub repository page](https://github.com/manabe0322/relsearch/releases).

5. Install ”relsearch_1.0.0.zip” from ”Install package(s) from local files...”.

\newpage

# Quick guide

1. Execute the following commands in R to start GUI.

&nbsp;

```{r eval = FALSE}
library(relsearch)
relsearch()
```

&nbsp;

2. Load files from each **"Browse..."** button.

![](fig_quick_1.jpg)

**Note** See section [Files](#sec_files) for information on each file.

\newpage

3. Click the **"Analysis"** button.

![](fig_quick_2.jpg)

\newpage

4. After finishing the analysis, the result window is automatically opened.

![](fig_quick_3.jpg)

**Note** See section [Result](#sec_result) for navigation of the result window.

\newpage

5. Select **Project** $\blacktriangleright$ **Save project**.

![](fig_quick_4.jpg)

\newpage

6. Enter the project name and click the **"Save as"** button.

![](fig_quick_5.jpg)

**Note** The saved project can be loaded from **Project** $\blacktriangleright$ **Load project**. How to handle projects is described in section [Other functions](#sec_other_func).

\newpage

# Files {#sec_files}

## STR: Victim database

```{r, echo = FALSE}
df_v_auto_ex <- data.frame(SampleName = c("Victim1", "Victim2", "Victim3", "Victim4", "Victim5", "Victim6", "Victim7", "Victim8", "Victim9", "Victim10", "︙"), 
                           D3S1358 = c("17", "17", "15", "15", "15", "16", "14", "16", "15", "15", "︙"), 
                           D3S1358 = c("17", "", "17", "18", "15", "16", "15", "18", "15", "16", "︙"), 
                           vWA = c("17", "18", "17", "15", "18", "17", "17", "17", "16", "18", "︙"), 
                           vWA = c("", "18", "18", "18", "", "", "18", "18", "18", "19", "︙"), 
                           D16S539 = c("11", "", "10", "9", "9", "", "12", "11", "11", "9", "︙"), 
                           D16S539 = c("", "", "10", "12", "9", "", "", "", "11", "10", "︙"), 
                           CSF1PO = c("12", "10", "10", "11", "10", "10", "", "9", "12", "12", "︙"), 
                           CSF1PO = c("", "11", "12", "12", "", "13", "", "12", "12", "12", "︙"), 
                           TPOX = c("", "8", "8", "8", "9", "8", "8", "8", "8", "8", "︙"), 
                           TPOX = c("", "", "11", "8", "11", "8", "8", "", "11", "11", "︙"))

kable(df_v_auto_ex, format = "latex", booktabs = T, label = NA) %>%
  kable_styling(latex_options = "HOLD_position", font_size = 10)
```

**Note**

- File type: .csv

- This file requires the column "SampleName" and columns for each marker (two columns in each).

- The marker with two empty cells (e.g., TPOX of sample "Victim1" in the above table) is ignored when calculating the likelihood ratio.

- The marker with one empty cell (e.g., vWA of sample "Victim1" in the above table) can be regarded as both homozygote (i.e., no drop-out) and heterozygote with drop-out of one allele when calculating the likelihood ratio.

&nbsp;

## STR: Reference database

```{r, echo = FALSE}
df_r_auto_ex <- data.frame(SampleName = c("Reference1", "Reference2", "Reference3", "Reference4", "Reference5", "Reference6", "Reference7", "Reference7", "Reference8", "Reference9", "Reference10", "︙"), 
                           Relationship = c("Father_Son", "Father_Son", "Father_Son", "Father_Son", "Brother_Brother", "Brother_Brother", "Brother_Brother", "Father_Son", "Brother_Brother", "Uncle_Nephew", "Uncle_Nephew", "︙"),
                           D3S1358 = c("17", "17", "15", "15", "15", "16", "16", "16", "15", "15", "15", "︙"), 
                           D3S1358 = c("17", "18", "16", "16", "16", "17", "17", "17", "16", "15", "16", "︙"), 
                           vWA = c("14", "17", "17", "15", "15", "17", "14", "14", "17", "14", "17", "︙"), 
                           vWA = c("17", "18", "18", "18", "18", "17", "18", "18", "18", "16", "", "︙"), 
                           D16S539 = c("10", "12", "9", "9", "9", "10", "9", "9", "9", "10", "9", "︙"), 
                           D16S539 = c("10", "13", "10", "9", "9", "12", "9", "9", "12", "11", "12", "︙"), 
                           CSF1PO = c("12", "10", "11", "10", "10", "13", "12", "12", "11", "11", "", "︙"), 
                           CSF1PO = c("12", "11", "12", "12", "12", "13", "15", "15", "12", "12", "", "︙"), 
                           TPOX = c("8", "11", "8", "9", "9", "9", "8", "8", "8", "8", "8", "︙"), 
                           TPOX = c("8", "11", "8", "11", "11", "11", "8", "8", "9", "9", "12", "︙"))
```

**Note**

- File type: .csv

- This file requires the columns "SampleName", "Relationship", and columns for each marker (two columns in each).

- The names of the relationship need to be pre-defined in **Settings** ▶ **Relationships** (see section [Setting](#sec_setting_rel)).

- If a reference has multiple missing family members, add rows for each relationship of the members (e.g., sibling and parent-child of the sample "Reference7" in the above table).

- The marker with two empty cells (e.g., CSF1PO of sample "Reference10" in the above table) is ignored when calculating the likelihood ratio.

- The marker with one empty cell (e.g., vWA of sample "Reference10" in the above table) can be regarded as both homozygote (i.e., no drop-out) and heterozygote with drop-out of one allele when calculating the likelihood ratio.

## STR: Allele frequencies

```{r, echo = FALSE}
df_af_ex <- data.frame(Allele = c("10", "10.1", "10.2", "10.3", "11", "11.1", "11.2", "11.3", "12", "12.2", "13", "13.2", "14", "︙"), 
                       D3S1358 = c("", "", "", "", "", "", "", "", "0.02324", "", "0.001328", "", "0.02656", "︙"), 
                       vWA = c("", "", "", "", "", "", "", "", "", "", "0.000664", "", "0.194887", "︙"), 
                       D16S539 = c("0.200266", "", "", "", "0.186981", "", "", "", "0.172036", "", "0.069744", "", "0.008967", "︙"), 
                       CSF1PO = c("0.215471", "", "", "", "0.206175", "", "", "", "0.420983", "", "0.069389", "", "0.017596", "︙"), 
                       TPOX = c("0.033566", "", "", "", "0.356597", "", "", "", "0.038219", "", "0.001329", "", "0.000997", "︙"))
```

**Note**

- File type: .csv

- This file requires the columns "Allele" and columns for each marker (one column in each).

&nbsp;

## Y-STR: Victim database

**Note**

- File type: .csv

- This file requires the column "SampleName" and columns for each marker (one column in each).

- In the marker with more than one allele, each allele must be separated by a comma without any spaces (e.g., DYS385).

- The marker with an empty cell (e.g., DYS456 of sample ‘Victim6’) is ignored for analysis.

&nbsp;

## Y-STR: Reference database

**Note**

- File type: .csv

- This file requires the column "SampleName", "Relationship", and columns for each marker (one column in each).

- The relationship of a missing family member should be designated in the column "Relationship". The name of the relationship should be defined in Settings > Relationships.

- If a reference has multiple missing family members, add rows for each relationship of the members (e.g., sibling and parent-child of the sample "Reference7" in the above table).

- In the marker with more than one allele, each allele must be separated by a comma without any spaces (e.g., DYS385).

- The marker with an empty cell (e.g., DYS518 of sample ‘Reference10’) is ignored for analysis.

&nbsp;

## mtDNA: Victim database

**Note**

- File type: .csv

- This file requires the column "SampleName", "Range", and "Haplotype".

&nbsp;

## mtDNA: Reference database

**Note**

- File type: .csv

- This file requires the column "SampleName", "Relationship", "Range", and "Haplotype".

- The relationship of a missing family member should be designated in the column "Relationship". The name of the relationship should be defined in Settings > Relationships.

- If a reference has multiple missing family members, add rows for each relationship of the members (e.g., sibling and parent-child of the sample "Reference7" in the above table).

&nbsp;

\newpage

# Setting {#sec_setting}

## Criteria

You can set the following criteria to support the assumed relationship. Press the **Save** button to reflect the changes. Press the **Reset** button if you want to return to default values.

### STR

- Minimum LR: The minimum likelihood ratio that supports the assumed relationship.

### Y-STR

- Maximum number of mismatched loci: The upper limit of the number of mismatched loci between victim and reference profiles to support paternal lineage.

- Maximum total mutational steps: The upper limit of total mutational steps (i.e., sum of mutational steps of all loci) between victim and reference profiles to support paternal lineage.

### mtDNA

- Maximum number of inconsistency: The upper limit of inconcistency between victim and reference profiles to support maternal lineage.

&nbsp;

## Relationship {#sec_setting_rel}

The names of the relationship in each reference profile need to be pre-defined. There are 44 pre-defined victim-reference relationships as the default.

Information on the table is as follows.

- **Pr(IBD = 2)**: The probability that two alleles are identity by descent (IBD).

- **Pr(IBD = 1)**: The probability that one allele is identity by descent (IBD).

- **Pr(IBD = 0)**: The probability that zero allele is identity by descent (IBD).

- **Paternal lineage**: Information on whether each relationship is paternal lineage or not.

- **Maternal lineage**: Information on whether each relationship is maternal lineage or not.

You can edit, add, and delete the relationships from the left sidebar.

### Edit

You can edit the defined relationship name from the **Edit** button.

In the pop-up window, select a relationship from the defined relationships, and enter a new name of the selected relationship. Press the **Save** button to reflect the new name.

### Add

You can define a new relationship from the **Add** button.

In the pop-up window, you have to set the relationship name and make a family tree to define the relationship between a victim and a reference.

The following is a list of functions to make a family tree.

- **Add a person**: The function to add an unknown person (UK) to the list of persons.

- **Delete a person**: The function to delete an unknown person (UK) from the list of person.

- **Sex**: The function to select a biological sex (male: M or female: F).

- **Father**: The function to select a person who is the father.

- **Mother**: The function to select a person who is the mother.

- **Founder**: The function to select whether the person is a founder of the family tree or not.

After defining a family tree, press the **View family tree** button to check whether the family tree is correctly set up or not.

If there is no problem in the family tree, press the **Save** button.

**Note** Information on the IBD probabilities, paternal lineage, and maternal lineage in the new relationship is automatically determined based on your-defined family tree.

### Delete

You can delete a defined relationship name from the **Delete** button.

In the pop-up window, select a relationship from the defined relationships. Press the **Save** button to delete the selected relationship.

### Reset

If you want to return to default settings, press the **Reset** button. In the pop-up window, press the **Restore default** button.

### Family tree

You can check the family tree of each defined relationship. Select a row from the relationship table and press the **Family tree** button.

## Mutation rate {#sec_setting_myu}

Mutation rates of each autosomal STR marker need to be set to calculate the likelihood ratio (LR). The default mutation rates considering the difference between paternal and maternal and mutational steps (-1, +1, -2, and +2) are reported by Morimoto et al [@Morimoto2020]. You can edit, add, and delete these mutation rates from the left sidebar.

### Edit

You can edit the mutation rates from the **Edit** button.

In the pop-up window, select a locus and change the mutation rates. Press the **Save** button to reflect the changes.

### Add

You can add a locus from the **Add** button.

In the pop-up window, enter a locus name and enter the mutation rates. Press the **Save** button to add the locus.

### Delete

You can delete a locus from the **Delete** button.

In the pop-up window, select a locus. Press the **Save** button to delete the selected locus.

### Reset

If you want to return to default settings, press the **Reset** button. In the pop-up window, press the **Restore default** button.

## Other parameters {#sec_setting_param}

You can set the minimum allele frequency which is used for the unobserved alleles. Press the **Save** button to reflect the changes. Press the **Reset** button if you want to return to the default value.

\newpage

# Result {#sec_result}

## Summary

After finishing the analysis, summary data is automatically displayed. You can change the displayed data from the left sidebar.

### Default display

Default display depends on whether the input data includes autosomal STR data or not.

If the input data includes autosomal STR data, the data where the LR exceeds the minimum LR (see section [Setting](#sec_setting)) is displayed.

If the input data is composed of only Y-STR and/or mtDNA data, the data that satisfies the criteria to support paternal/maternal lineage (see section [Setting](#sec_setting)) is displayed.

If the number of data that satisfies the criteria is more than 10,000, the top 10,000 data is displayed.

### Identified pairs

If the data supports that a victim (or a reference) has the assumed relationship with only one reference (or victim), the victim-reference pair is categorized as the identified pair.

### Multiple candidates

If the data supports that a victim (or a reference) has the assumed relationship with multiple references (or victims), the victim-reference pair is categorized as the multiple candidates.

### Not support lineage

If the LR exceeds the minimum LR but the data does not satisfy the criteria to support paternal/maternal lineage when assuming the paternal/maternal lineage relative, the victim-reference pair is categorized as not support lineage.

### Minimum LR displayed

You can change the minimum LR displayed. The lower limit is LR = 1. If the number of data where the LR exceeds the minimum LR is more than 10,000, the top 10,000 data is displayed.

### Download summary data

You can download summary data as a .csv file. Press the **Downlaod** button in Summary tab.

## Selected data in detail

The data of a selected victim-reference pair is displayed in **Selected data in detail** tab.

1. Select a row in the summary table.

2. Go to **Selected data in detail** tab.

In **Selected data in detail** tab, you can check and download the analyzed data of STR, Y-STR, and mtDNA in the selected victim-reference pair. 

## Analysis conditions

You can check the following analysis conditions in **Analysis conditions** tab.

- **Database**: Input .csv file names of each database are displayed.

- **Allele probability**: Unobserved alleles in the population database are displayed, if any. The frequency of these alleles is set to the user-defined minimum allele frequency and the sum of the frequencies in each locus is corrected to 1.

- **Criteria**: The user-defined criteria to support the assumed relationship (see section [Setting](#sec_setting)) is displayed.

- **Assumed relationship**: Information on the user-defined relationships (see section [Setting](#sec_setting_rel)) is displayed.

- **Mutation rate**: The user-defined mutation rates (see section [Setting](#sec_setting_myu)) are displayed.

- **Parameter**: The user-defined minimum allele frequency (see section [Setting](#sec_setting_param)) is displayed.

\newpage

# Other functions {#sec_other_func}

## Project

### New project

To start a new project, go to **Project** $\blacktriangleright$ **New project** and press the **New project** button.

**Important**: The current project should be saved before starting a new project because the current project will be deleted.

### Load project

To load a previous project, go to **Project** $\blacktriangleright$ **Load project**. Select a project file from the **Browse...** button and press the **Load project** button.

**Important**: The current project should be saved before loading another project because the current project will be deleted.

### Save project

To save the current project, go to **Project** $\blacktriangleright$ **Save project**. Enter the project name and press the **Save as** button.

## View database

Each loaded database can be checked in **Database** tab.

## Example files

Example files for each database can be downloaded from the **Example files** tab.

\newpage

# Computational principle

## STR

### Likelihood ratio

The likelihood ratio (LR) is calculated for each victim-reference pair by assuming the following two hypotheses:

$H_{1}$: The victim and the reference are a certain relationship,

$H_{2}$: The victim and the reference are unrelated.

The equation of the LR is as follows:

$$
LR = \frac{Pr(G_{v}, G_{r}|H_{1})}{Pr(G_{v}, G_{r}|H_{2})}
$$

where $G_{v}$ and $G_{r}$ denote the genotypes of the victim and the reference, respectively.

In RelSearch, users can set the minimum threshold of the LR to support $H_{1}$. The default threshold is 100, which is determined based on the verbal scale "Moderate Support" written in the Scientific Working Group on DNA Analysis Methods (SWGDAM) guideline [@SWGDAM2018].

If full profiles are obtained for a victim-reference pair under the assumption of the relationship other than the parent-child (i.e., without considering drop-out and mutations in calculation), the numerator and the denominator of the LR are calculated using the probabilities that two, one and zero alleles are identity by descent (IBD).

$$
\begin{aligned}
Pr(G_{v}, G_{r}|H) &= Pr(G_{v}, G_{r}|IBD = 2)Pr(IBD = 2|H)
\\ &+ Pr(G_{v}, G_{r}|IBD = 1)Pr(IBD = 1|H)
\\ &+ Pr(G_{v}, G_{r}|IBD = 0)Pr(IBD = 0|H)
\end{aligned}
$$

The specific method for these calculations is based on the method of Wenk et al [@Wenk1996].

### Allele probabilities

The allele probabilities are assumed to be based on Dirichlet distributions in each locus [@Morimoto2020]. The probability of allele $x$ ($p_x$) can be estimated using the following formula:

$$
p_a = \frac{\alpha_x + 1}{\sum_{x = 1}^X(\alpha_x + 1)}
$$

where $\alpha_x$ is the number of observations of the $x$th allele ($x = 1, 2, ..., X$) and X is the number of observed allele types.

If some alleles of victims or references are unobserved in the population database, these alleles are added to the observed allele types as zero observation (i.e., $\alpha_x = 0$), and $p_x$ of all alleles are re-estimated [@Morimoto2020].

### Allelic drop-out

When only one allele is observed and another allele may have dropped out in a locus, RelSearch assumes that the genotype is both homozygous (i.e., no drop-out allele) and heterozygous (i.e., one drop-out allele). The probability of observing the profile is the same in all candidate genotypes. The method corresponds to the "Method 1" proposed by Dørum et al [@Dorum2015].

Here, $G_{v}^*$ and $G_{r}^*$ denote the observed profiles of the victim and the reference, respectively. $G_{v_i}$ and $G_{r_j}$ denote the $i$th candidate genotype of the victim and the $j$th candidate genotype of the reference, respectively.

Suppose that $G_{v}^*$ is 15/- (i.e., partial profile) and $G_{r}^*$ is 16/17. In this case, RelSearch considers the following candidate genotypes of the victim:

$G_{v_1}$: 15/15, 

$G_{v_2}$: 15/16, 

$G_{v_3}$: 15/17, and

$G_{v_4}$: 15/Q.

Q denotes any undetected alleles other than 15, 16, and 17.

The candidate genotype of the reference is only 16, 17 because two heterozygote alleles are detected.

In this case, the likelihood function considering allelic drop-out is as follows:

$$
\begin{aligned}
\\ &Pr(G_{v}^* = 15/-, G_{r}^* = 16/17|H)
\\ &= Pr(G_{v}^* = 15/-, G_{r}^* = 16/17|G_{v_i} = 15/15, G_{r} = 16/17)Pr(G_{v_i} = 15/15, G_{r} = 16/17|H)
\\ &+ Pr(G_{v}^* = 15/-, G_{r}^* = 16/17|G_{v_i} = 15/16, G_{r} = 16/17)Pr(G_{v_i} = 15/16, G_{r} = 16/17|H)
\\ &+ Pr(G_{v}^* = 15/-, G_{r}^* = 16/17|G_{v_i} = 15/17, G_{r} = 16/17)Pr(G_{v_i} = 15/17, G_{r} = 16/17|H)
\\ &+ Pr(G_{v}^* = 15/-, G_{r}^* = 16/17|G_{v_i} = 15/Q, G_{r} = 16/17)Pr(G_{v_i} = 15/Q, G_{r} = 16/17|H)
\end{aligned}
$$

This equation is generalized as follows:

$$
\begin{aligned}
Pr(G_{v}^*, G_{r}^*|H) &= \sum_i \sum_j Pr(G_{v}^*, G_{r}^*|G_{v_i}, G_{r_j})Pr(G_{v_i}, G_{r_j}|H)
\end{aligned}
$$

$Pr(G_{v}^*, G_{r}^*|G_{v_i}, G_{r_j})$ can be ignored because the probability of observing $G_{v}^*$ and $G_{r}^*$ is the same irrespective of the candidate genotypes. Therefore, the likelihood function can be rewritten as follows:

$$
\begin{aligned}
Pr(G_{v}^*, G_{r}^*|H) &= \sum_i \sum_j Pr(G_{v_i}, G_{r_j}|H)
\end{aligned}
$$

$Pr(G_{v_i}, G_{r_j}|H)$ values are calculated according to equation (1) under the assumption of the relationship other than the parent-child. When assuming the parent-child relationship, mutational events are considered for the calculation as explained in the next session.

### Mutation

When assuming the parent-child relationship, mutational events are considered for the calculation of $Pr(G_{v_i}, G_{r_j}|H)$ in equation (2). Suppose that a victim-reference pair is assumed to be parent-child, and $G_{v_i}$ is a/b and $G_{r_j}$ is c/d. RelSearch assumes all inheritance patterns from the parent to the child including mutational events. The LR is calculated according to [@Ricciardi2014; @Egeland2017].

$$
LR = \frac {(\mu_{a \rightarrow c} + \mu_{b \rightarrow c})p_{d} + (\mu_{a \rightarrow d} + \mu_{b \rightarrow d})p_{c}}{4p_{c}p_{d}}
$$

where $p_a$, $p_b$, $p_c$, and $p_d$ denote the probabilities of the allele $a$, $b$, $c$, and $d$, respectively. $\mu_{a \rightarrow c}$ denotes the probability that the parent allele $a$ is inherited as the allele $c$ to the child. $\mu_{a \rightarrow c}$, $\mu_{b \rightarrow c}$, $\mu_{a \rightarrow d}$, and $\mu_{b \rightarrow d}$ depend on the sex of the parent (i.e., paternal or maternal), and the mutational steps are assumed to be -2, -1, +1, and +2 steps. The mutation rates reported by Morimoto et al [@Morimoto2020] are used as the default values.

## Y-STR

RelSearch analyzes whether the Y-STR profiles of each victim-reference pair support paternal lineage. There are two criteria to support paternal lineage as follows:

- Maximum number of mismatched loci: The upper limit of the number of mismatched loci between victim and reference profiles to support paternal lineage.

- Maximum total mutational steps: The upper limit of total mutational steps (i.e., sum of mutational steps of all loci) between victim and reference profiles to support paternal lineage.

If a locus has no alleles in at least one of the victim and the reference, the locus is ignored for the analysis. In addition, if one of the profiles of victim and the reference is partial in a duplicated marker (e.g., DYS385), and the profiles can be explained as paternal lineage (e.g., victim profile: 15, reference profile: 15/16), the locus is also ignored for the analysis.

## mtDNA

RelSearch analyzes whether the mtDNA profiles of each victim-reference pair support maternal lineage. There is one criterion to support maternal lineage as follow:

- Maximum number of inconsistency: The upper limit of the number of inconcistency between victim and reference profiles to support maternal lineage.

Information on the ranges of the mtDNA sequences in each profile is needed because RelSearch targets the ranges read in both victim and reference profiles. The mtDNA profiles are expressed as the differences to a reference sequence such as the revised Cambridge Reference Sequence [@Andrews1999] (e.g., 73G, 315.1C, and so on). RelSearch only recongizes differences in strings of each mtDNA profile; therefore, the rule of the notation for mtDNA profiles should be aligned in each profile by software users.

The default criterion has been determined according to the sensitivity and specificity studies as explained in section 2.5.3.

\newpage

# Reference {-}

<div id="refs"></div>
